{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1acb1e0f",
   "metadata": {},
   "source": [
    "# Usage of EC for the Traveling Salesman Problem\n",
    "## CI2025_Lab2\n",
    "**K/BIDI Thomas**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85ad858e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab78a9d5",
   "metadata": {},
   "source": [
    "## Problems\n",
    "\n",
    "We are trying to solve the Traveling Salesman Problem (and some variants of it) using **Evolutionnary Computation algorithms**.  \n",
    "To achieve either an optimal or \"good enough\" solution for each problems we're gonna explore an implementation of the EC with the following definitions and properties : \n",
    "(We're using the terms of the classic TSP, which consists of finding a minimal-cost hamiltonian cycle in an undirected complete graph, where each edge's weight is the distance between the two cities it connects. For the R1 and R2 variants, the ideas are the same, except the concept of distance cannot be applied to them [matrix are not always symmetric, triangular inequality does not necessary hold].)\n",
    "\n",
    "- **Fitness** : The fitness represent the sum of distances between the cities of the solution (which includes the distance between the final and first cities)\n",
    "- **Individual** : An individual is a sequence of cities describing the hamiltonian cycle, it holds this properties $$|i| = |V|$$ with  $i$ an individual, and $V$ the set of vertices of the graph. So in our case, a gene is a city of the cycle.\n",
    "- **Population** : The population is a set of individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e756b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here was my first implementation of the fitness\n",
    "def fitness(solution, problem):\n",
    "    result = problem[solution[-1], solution[0]]\n",
    "    for i in range(len(solution) - 1):\n",
    "        result += problem[solution[i], solution[i+1]]\n",
    "    return result\n",
    "\n",
    "def population_fitness(population, problem):\n",
    "    # Return the fitness for every individuals in the population\n",
    "    pop_fitness = [fitness(i, problem) for i in population]\n",
    "    return pop_fitness"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6577925d",
   "metadata": {},
   "source": [
    "We're going to implement various functions, the idea is to explore the fitness space and find a pretty good, if not *optimal*, solution.  \n",
    "\n",
    "Firstly, let's define the main function we're going to use, this function will create the initial population and realize the different steps of EC :\n",
    "- Initialize the population randomly\n",
    "- Realize the selection of the parents\n",
    "- Realize the crossovers and mutations\n",
    "- Select the offspring \n",
    "- Keep track of the fitness over each generations and plot a graph of it at the end\n",
    "- Keep track of the overall best solutions and fitness and return them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db1f595",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function used for early stopping, here we will stop when the generation best fitness do not move during (10% * N_GENS) consecutives generations.\n",
    "def on_plateau_for_K(fitness_list, K):\n",
    "    if (len(fitness_list) < K):\n",
    "        return False\n",
    "    last_K = fitness_list[-K:]\n",
    "    return np.all(last_K == last_K[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "496ad380",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generation(problem, N_GENS, POP_SIZE, parent_selection, survivor_selection, offsprings_generation, mutation, mu_coeff, l_coeff, problem_name=\"\"):\n",
    "    \"\"\"\n",
    "    problem : the matrix of distances between the cities\n",
    "    N_GENS : the number of generations on which the evolutionary process will be applied\n",
    "    POP_SIZE : the number of individuals in the original population\n",
    "    parent_selection : function used to select the parent individuals\n",
    "    offsprings_generation : function performing recombination and mutation on the parent population to generate the offspring\n",
    "    survivor_selection : function used to select the new population from the offspring\n",
    "    mutation : function applied during offspring generation\n",
    "    mu_coeff : gives mu (the portion of the population selected as parents) when multiplied by POP_SIZE\n",
    "    l_coeff : gives lambda (the offspring and new population size) when multiplied by POP_SIZE\n",
    "    problem_name : name of the problem to display on the generated graph\n",
    "    \"\"\"\n",
    "    # Random population generation\n",
    "    population = np.array([np.random.permutation(np.arange(problem.shape[0])) for _ in range(POP_SIZE)])\n",
    "    \n",
    "    mu = int(POP_SIZE * mu_coeff)\n",
    "    l = int(POP_SIZE * l_coeff)\n",
    "\n",
    "    # Keep track of the best individual in all generations\n",
    "    best_individual = population[0]\n",
    "    best_fitness = fitness(best_individual, problem)\n",
    "    best_fitness_generation = -1 # Generation in which the best_fitness appeared\n",
    "\n",
    "    # Keep tracks of all the generations best fitness to plot the graph of its evolution\n",
    "    fitness_list = []\n",
    "    plateau_size = int(N_GENS * 0.1) # 10%\n",
    "    num_gens = N_GENS # Value to keep plot only on the actual number of generations made\n",
    "\n",
    "    # Use of TQDM to show a progress bar\n",
    "    for gen in tqdm(range(N_GENS), desc=\"Evolution Computation\", unit=\"gen\"):\n",
    "        # Usage of the given function to realize the different steps of the EC\n",
    "        parents = parent_selection(population, mu, l, problem)\n",
    "        offsprings = offsprings_generation(parents, l, problem, mutation)\n",
    "        survivors = survivor_selection(parents, offsprings, mu, l)\n",
    "        population = survivors\n",
    "\n",
    "        generation_best_fitness = 1e30 # Suppose to be more that the worst possible fitness\n",
    "        \n",
    "        fitness_by_individual = population_fitness(population, problem)\n",
    "        sort_indices = np.argsort(fitness_by_individual)\n",
    "        generation_best_fitness = fitness_by_individual[sort_indices[0]]\n",
    "        if (generation_best_fitness < best_fitness):\n",
    "            best_individual = population[sort_indices[0]]\n",
    "            best_fitness = generation_best_fitness\n",
    "            best_fitness_generation = gen + 1\n",
    "        \n",
    "        fitness_list.append(generation_best_fitness)\n",
    "        if (on_plateau_for_K(fitness_list, plateau_size)):\n",
    "            print(f\"Early stopping due to best generation fitness not moving for {plateau_size} generations\")\n",
    "            num_gens = gen + 1\n",
    "            break\n",
    "\n",
    "    print(f\"The best fitness is {best_fitness} and it was found in generation : {best_fitness_generation}\")\n",
    "    # We plot the graph of the fitness over the generations\n",
    "    plt.plot(range(num_gens), fitness_list)\n",
    "    # We get the first value closer to the best fitness and add a point here\n",
    "    plt.scatter(best_fitness_generation, best_fitness, color=\"red\", label=\"Best fitness\")\n",
    "    plt.xlabel(\"Generations\")\n",
    "    plt.ylabel(\"Fitness\")\n",
    "    plt.title(f\"Evolution of fitness over generations ({problem_name})\")\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "\n",
    "    return best_individual, best_fitness"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21da5b54",
   "metadata": {},
   "source": [
    "### Parent selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66c06720",
   "metadata": {},
   "source": [
    "For the parent selection, multiple options exists and we are going to implement 2 of them :\n",
    "- **Uniform selection** : we randomly pick one the parents from the population\n",
    "- **Best selection** : we sort the fitness values of all the individuals in the population and we select the $\\mu$ first as the parents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d6c2e2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uniform_selection(population, mu, l, problem):\n",
    "    # We select mu individuals in the population randomly\n",
    "    selection = np.random.choice(len(population), size=mu, replace=False)\n",
    "    parents = np.array([population[i] for i in selection], dtype=int)\n",
    "    return parents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70c69b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_selection(population, mu, l, problem):\n",
    "    # We select the mu best individuals in the population\n",
    "    fitness_by_individual = population_fitness(population, problem)\n",
    "    sort_indices = np.argsort(fitness_by_individual)\n",
    "    parents = np.array([population[sort_indices[i]] for i in range(mu)])\n",
    "    return parents"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edf24530",
   "metadata": {},
   "source": [
    "### Crossover\n",
    "\n",
    "For our problem we need a crossover method which can preserve the validity of the individuals, this condition make the n-point crossover a bad option, since it can generate individuals which are not an hamiltonian cycle.  \n",
    "To answer this problem we're going to use the **Order Crossover (OX)**, the idea behind this crossover is that we select a subsequence of the first parent and we copy it at the same exact position inside the genotype of the offspring, then we complete the remaining empty genes with those of the second parent which are not already in the offspring's genotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39136624",
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_crossover(parent1, parent2):\n",
    "    # Choice of the start and end of the substring inside parent1\n",
    "    start, end = np.sort(np.random.choice(len(parent1), size=2, replace=False))\n",
    "    # Creation of the child\n",
    "    child = [-1 for _ in range(len(parent1))]\n",
    "    child[start:end] = parent1[start:end] # Copy of the subsequence in it\n",
    "\n",
    "    # We create a mask and remove the already existing genes of the offsprings from the parent2\n",
    "    mask = ~np.isin(parent2, child)\n",
    "    parent2_filtered = parent2[mask]\n",
    "    starting_index = 0 if start > 0 else end\n",
    "\n",
    "    # We copy the remaining genes in the available place of the offspring's genotype\n",
    "    for i in parent2_filtered:\n",
    "        child[starting_index] = i\n",
    "        starting_index += 1\n",
    "        if (starting_index >= start and starting_index < end):\n",
    "            starting_index = end\n",
    "\n",
    "    return child"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9170ff64",
   "metadata": {},
   "source": [
    "### Mutations\n",
    "\n",
    "Every once in a while it is interesting for an offspring to mutate in order to favor the exploration of the fitness space. We are going to implement various simple mutation functions, which are the following ones : \n",
    "- **Swap mutation** : We randomly select 2 genes and we swap their positions.\n",
    "- **Scramble mutation** : We randomly select 2 sequences and we swap their positions.\n",
    "- **Insert mutation** : We randomly select a gene and move it.\n",
    "- **Inversion mutation** : We randomly select a sequence in the individual and inverse it.\n",
    "- **Random mutation** : We pick uniformely one of the previous mutation and apply it to the individual."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d2d4b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Swap mutation\n",
    "def swap_mutation(individual):\n",
    "    # Swap 2 elements in the individual\n",
    "    gene1, gene2 = np.sort(np.random.choice(len(individual), size=2, replace=False))\n",
    "    \n",
    "    temp = individual[gene1]\n",
    "    individual[gene1] = individual[gene2]\n",
    "    individual[gene2] = temp\n",
    "\n",
    "    return individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7940e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scramble mutation\n",
    "def scramble_mutation(individual):\n",
    "    # Swap 2 sequences in the individual\n",
    "    start1, end1, start2, end2 = np.sort(np.random.choice(len(individual), size=4, replace=False))\n",
    "    \n",
    "    # Retrieve the sequences\n",
    "    segment1 = individual[start1:end1]\n",
    "    segment2 = individual[start2:end2]\n",
    "\n",
    "    new_individual = np.concatenate([\n",
    "        individual[:start1],\n",
    "        segment2,\n",
    "        individual[end1:start2],\n",
    "        segment1,\n",
    "        individual[end2:]\n",
    "    ]).astype(int)\n",
    "\n",
    "    return new_individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341db342",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert mutation\n",
    "def insert_mutation(individual):\n",
    "    # Randomly select the position of the gene to move\n",
    "    gene_position, new_position = np.random.choice(len(individual), size=2, replace=False)\n",
    "    \n",
    "    # We delete and re-instert the gene\n",
    "    gene = individual[gene_position]\n",
    "    new_individual = np.delete(individual, gene_position)\n",
    "    new_individual = np.insert(new_individual, new_position, gene)\n",
    "\n",
    "    return new_individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8247d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inversion mutation\n",
    "def inversion_mutation(individual):\n",
    "    start, end = np.sort(np.random.choice(len(individual), size=2, replace=False))\n",
    "    \n",
    "    # We retrieve and inverse the segment\n",
    "    segment = individual[start:end]\n",
    "    inversed_segment = segment[::-1]\n",
    "\n",
    "    # We concatenate the untouched parts with the reversed segments\n",
    "    new_individual = np.concatenate([\n",
    "        individual[:start],\n",
    "        inversed_segment,\n",
    "        individual[end:]\n",
    "    ]).astype(int)\n",
    "\n",
    "    return new_individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d8d723b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uniform selection in a set of mutations (swap, scramble, insert, inversion)\n",
    "def random_mutation(individual):\n",
    "    mutations = [swap_mutation, scramble_mutation, insert_mutation, inversion_mutation]\n",
    "    choice = np.random.choice(len(mutations))\n",
    "    return mutations[choice](individual)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5692967f",
   "metadata": {},
   "source": [
    "### Offsprings generation\n",
    "\n",
    "We need a function to generate the offsprings from the parents. We are going to select **uniformely** 2 parents in the given array and use the order crossover to produce an offspring. A *mutation* could appear with a probability of 40%."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4947da25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def offsprings_generation(parents, l, problem, mutation):\n",
    "    # Generate lambda (l) offsprings from the parents population\n",
    "    offsprings = []\n",
    "\n",
    "    while len(offsprings) < l:\n",
    "        # Parents are already the best of the population, we uniformely pick inside them\n",
    "        p1, p2 = uniform_selection(parents, 2, -1, problem)\n",
    "        offspring = order_crossover(p1, p2)\n",
    "        if (np.random.rand() < 0.4): # Probability of 40% of a mutation\n",
    "            offspring = mutation(offspring)\n",
    "        offsprings.append(offspring)\n",
    "\n",
    "    return np.array(offsprings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bf4387c",
   "metadata": {},
   "source": [
    "### Survivor selection\n",
    "\n",
    "We need to decide of a policy for the selection of the survivors which are going to replace the population for the next generation, here we are going to use this one :\n",
    "- **Generational** : We replace all the population by the offsprings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cafbfa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generational_selection(parents, offsprings, mu, l):\n",
    "    return offsprings # Only the offsprings are selected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edfb2d74",
   "metadata": {},
   "source": [
    "### Problems list\n",
    "\n",
    "All the problems we need to solve are in the lab2 subfolder, we're going to load them in a list. This will let us run all algorithms in a row and observe the results afterwards."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abcd3082",
   "metadata": {},
   "outputs": [],
   "source": [
    "problems = os.listdir(\"lab2/\")\n",
    "problems_info = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f92206eb",
   "metadata": {},
   "source": [
    "### Upgrade\n",
    "\n",
    "I noticed that this implementation was a bit slow due to the computation of all the fitness values, we are going to use the Numpy's vectorization capabilities, which gives us access to the underlying C execution frame, to accelerate the **fitness** and **population_fitness** computation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad84e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Vectorized version of the fitness and population_fitness\n",
    "def fitness(solution, problem):\n",
    "    return np.sum(problem[solution, np.roll(solution, -1)])\n",
    "\n",
    "def population_fitness(population, problem):\n",
    "    return np.sum(problem[population, np.roll(population, -1, axis=1)], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c136c3a2",
   "metadata": {},
   "source": [
    "### Final Test\n",
    "\n",
    "We are now going to develop a small loop exploiting the list of problems and the differents function we developped earlier to search for the solutions of each problems.\n",
    "\n",
    "Tweaking the parameters helps giving the best results, for this test I am going to set **N_GENS** at **5000** and the **POP_SIZE** at **500**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d34670a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "problems_info = {}\n",
    "\n",
    "for problem_name in problems:\n",
    "    problem = np.load(f\"lab2/{problem_name}\")\n",
    "    best_solution, best_fitness = generation(problem, 5000, 500, best_selection, generational_selection, offsprings_generation, random_mutation, 0.05, 1, problem_name)\n",
    "    problems_info[problem_name] = (best_solution, best_fitness)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
